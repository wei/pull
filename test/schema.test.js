/* eslint-env node, jest */

const schema = require('../lib/schema').schemaWithDeprecation

const validConfigs = [
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master' }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', autoMerge: true }], label: 'pull' }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', autoMerge: true, assignees: ['wei'] }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', autoMerge: false, reviewers: ['wei'] }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', autoMerge: false, reviewers: ['wei'], conflictReviewers: ['saurabh702'] }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', mergeMethod: 'squash', mergeUnstable: true }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', mergeMethod: 'hardreset', assignees: ['wei'] }] }],
  [{
    version: '1',
    rules: [
      { base: 'master', upstream: 'upstream:master', autoMerge: true },
      { base: 'development', upstream: 'upstream:development', autoMerge: false, autoMergeHardReset: true }
    ]
  }],
  [{
    version: '1',
    rules: [
      { base: 'master', upstream: 'upstream:master', autoMerge: true, autoMergeHardReset: true, assignees: ['wei'] },
      { base: 'development', upstream: 'upstream:development', autoMerge: false, autoMergeHardReset: true, reviewers: ['wei'] },
      { base: 'development', upstream: 'upstream:development', autoMerge: false, autoMergeHardReset: true, reviewers: ['wei'], conflictReviewers: ['saurabh702'] }
    ],
    label: 'pull'
  }]
]

const invalidConfigs = [
  { },
  { rules: {} },
  { version: '' },
  { version: '1' },
  { version: '1', rules: [] },
  { version: '1', rules: [{ base: 'master' }] },
  { version: 1, rules: [{ base: 'master', upstream: 'upstream:master' }] },
  { version: '1', rules: [{ base: 'master', upstream: 'upstream:master' }], label: 1 },
  { version: '1', rules: [{ base: 'master', upstream: 'upstream:master' }], label: '' },
  { version: '1', rules: [{ base: 'master', upstream: 'upstream:master', assignees: '' }] },
  { version: '1', rules: [{ base: 'master', upstream: 'upstream:master', reviewers: '' }] },
  { version: '1', rules: [{ base: 'master', upstream: 'upstream:master', reviewers: '', conflictReviewers: '' }] },
  { version: '1', rules: [{ base: 'master', upstream: '' }] },
  { version: '1', rules: [{ base: 'master', autoMerge: 1 }] },
  { version: '1', rules: [{ base: 'master', autoMerge: '' }] },
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', mergeMethod: '' }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', mergeMethod: 'invalid' }] }],
  [{ version: '1', rules: [{ base: 'master', upstream: 'upstream:master', mergeMethod: true }] }],
  {
    version: '1',
    rules: [{ base: 'master', upstream: 'upstream:master', autoMerge: true, autoMergeHardReset: 1 }]
  },
  {
    version: '1',
    rules: [{ base: 'master', upstream: 'upstream:master', autoMerge: true, autoMergeHardReset: '' }]
  }
]

describe('schema', () => {
  test('defaults', async () => {
    expect(schema.validate({ version: '1', rules: [{ base: 'master', upstream: 'upstream:master' }] }).value).toEqual({
      version: '1',
      rules: [
        {
          base: 'master',
          upstream: 'upstream:master',
          autoMerge: false,
          autoMergeHardReset: false,
          mergeMethod: 'none',
          mergeUnstable: false,
          assignees: [],
          reviewers: [],
          conflictReviewers: []
        }
      ],
      label: ':arrow_heading_down: pull'
    })
  })

  validConfigs.forEach(([example, expected = example]) => {
    test(`${JSON.stringify(example)} is valid`, () => {
      const { error, value } = schema.validate(example)
      expect(error).toBe(undefined)
      expect(value).toMatchObject(expected)
    })
  })

  invalidConfigs.forEach((example) => {
    test(`${JSON.stringify(example)} is invalid`, () => {
      const { error } = schema.validate(example)
      expect(error && error.toString()).toMatchSnapshot()
    })
  })
})
