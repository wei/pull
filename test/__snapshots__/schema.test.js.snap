// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema [{"version":"1","rules":[{"base":"master","upstream":"upstream:master","mergeMethod":""}]}] is invalid 1`] = `"ValidationError: \\"value\\" must be of type object"`;

exports[`schema [{"version":"1","rules":[{"base":"master","upstream":"upstream:master","mergeMethod":"invalid"}]}] is invalid 1`] = `"ValidationError: \\"value\\" must be of type object"`;

exports[`schema [{"version":"1","rules":[{"base":"master","upstream":"upstream:master","mergeMethod":true}]}] is invalid 1`] = `"ValidationError: \\"value\\" must be of type object"`;

exports[`schema {"rules":{}} is invalid 1`] = `"ValidationError: \\"version\\" is required"`;

exports[`schema {"version":""} is invalid 1`] = `"ValidationError: \\"version\\" is not allowed to be empty"`;

exports[`schema {"version":"1","rules":[]} is invalid 1`] = `"ValidationError: \\"rules\\" does not contain 1 required value(s)"`;

exports[`schema {"version":"1","rules":[{"base":"master","autoMerge":""}]} is invalid 1`] = `"ValidationError: \\"rules[0].upstream\\" is required"`;

exports[`schema {"version":"1","rules":[{"base":"master","autoMerge":1}]} is invalid 1`] = `"ValidationError: \\"rules[0].upstream\\" is required"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":""}]} is invalid 1`] = `"ValidationError: \\"rules[0].upstream\\" is not allowed to be empty"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master","assignees":""}]} is invalid 1`] = `"ValidationError: \\"rules[0].assignees\\" must be an array"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master","autoMerge":true,"autoMergeHardReset":""}]} is invalid 1`] = `"ValidationError: \\"rules[0].autoMergeHardReset\\" must be a boolean"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master","autoMerge":true,"autoMergeHardReset":1}]} is invalid 1`] = `"ValidationError: \\"rules[0].autoMergeHardReset\\" must be a boolean"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master","reviewers":"","conflictReviewers":""}]} is invalid 1`] = `"ValidationError: \\"rules[0].reviewers\\" must be an array"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master","reviewers":""}]} is invalid 1`] = `"ValidationError: \\"rules[0].reviewers\\" must be an array"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master"}],"label":""} is invalid 1`] = `"ValidationError: \\"label\\" is not allowed to be empty"`;

exports[`schema {"version":"1","rules":[{"base":"master","upstream":"upstream:master"}],"label":1} is invalid 1`] = `"ValidationError: \\"label\\" must be a string"`;

exports[`schema {"version":"1","rules":[{"base":"master"}]} is invalid 1`] = `"ValidationError: \\"rules[0].upstream\\" is required"`;

exports[`schema {"version":"1"} is invalid 1`] = `"ValidationError: \\"rules\\" is required"`;

exports[`schema {"version":1,"rules":[{"base":"master","upstream":"upstream:master"}]} is invalid 1`] = `"ValidationError: \\"version\\" must be a string"`;

exports[`schema {} is invalid 1`] = `"ValidationError: \\"version\\" is required"`;
